### PR TITLE
Create `haddock3-score` CLI

### PR DESCRIPTION
**Pull request by rvhonorato**
~_I deleted the PR template since this is an extra tool and not a module/gear/libs/core/clis/cns_~

~Adds a tool to calculate the haddock-score of a complex.~

~I see quite some more things that can be implemented such as support for pdb.list, custom-weights, use restraints, etc - we can open additional issues for that.~

~(venv) repos/haddock3 » pwd~
~/Users/rodrigo/repos/haddock3~
~(venv) repos/haddock3 » export PYTHONPATH=`pwd`/src~
~(venv) repos/haddock3 » time python tools/haddock-score.py examples/data/2oob.pdb
HADDOCK-score (emscoring): -53.8110~
~python tools/haddock-score.py examples/data/2oob.pdb  2.32s user 0.08s system 97% cpu 2.477 total~

@joaomcteixeira:

* Add a new command-line using the `libworkflow` machinery to run `emscoring` directly on a complex (PDB file) using the command line.
* Idea was develop by @rvhonorato and @amjjbonvin 
* @joaomcteixeira worked out the integration in haddock. See past commits for details.~

Cheers,

### CODE REVIEW

**Review by amjjbonvin — COMMENTED** [review]
Would it be an idea to add a few options to:

1) Possibly printout also the components of the score

2) Save the generated PDB file (now only kept into memory

This command should in principle return a score rather similar to the EM scoring example, right?

**Review by mgiulini — COMMENTED** [review]
User experience is great! but I agree that having the different components would be useful.

I calculated the haddock-score for the first 20 pdbs of 1_emscoring for capri Target218 (check out on tintin if you like). It is highly correlated to the true haddock score, but the pearson coeff. is not super (r~0.88). Should we add more steps to this? Or is it ok as it is?

**Comment by rvhonorato** [comment]
> The number of EM steps in emscoring in only 50 while it is 200 in emref And I assume autohis is true for both?

Yep, this tool runs `emscoring` under the hood

> And I assume autohis is true for both?

This tool uses the default parameters

**Comment by rvhonorato** [comment]
> * Possibly printout also the components of the score

Done

> * Save the generated PDB file (now only kept into memory

It is saved as `calc-hs_1.pdb` in the directory that you ran the command

**Comment by amjjbonvin** [comment]
> Possibly printout also the components of the score
> Done
> 
> 

As an option?

> Save the generated PDB file (now only kept into memoryI

I would rather save it as <original-pdb-filename)_hs.pdb  to avoid overwriting it if running on multiple PDBs in the same dir.

File: tools/haddock-score.py [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+        (1.0 * desolv) + (0.1 * air)
+
+    print("-----")
+    print(f"vdw\t{vdw:.4f}")
+    print(f"elec't{elec:.4f}")
+    print(f"desolv\t{desolv:.4f}")
+    print(f"air\t{air:.4f}")
+    print("HADDOCK-score = (1.0 * vdw) + (0.2 * elec) + (1.0 * desolv) + (0.1 * air)")
+    print("-----")
+    print(f"HADDOCK-score (emscoring) = {haddock_score_itw:.4f}")
```
Comment (amjjbonvin) [inline tools/haddock-score.py:193]: air will be in principle 0
may-be more interesting to output the BSA.
And I would not output everything by default, but only as an option (e.g. if the `-full` argument is given). 
Simpler for incorporating the score in other scripts.

Comment (amjjbonvin) [reply tools/haddock-score.py:193]: And same thing for outputting the `pdb` and `psf` files, only if an option is given, e.q. `-outpdb` -`outpsf`

**Comment by joaomcteixeira** [comment]
Hi,
Quick question before a more in-depth review. Why can't this be a CLI?
cheers

File: .github/workflows/tests.yml [hunk 2]
```diff
@@ -48,5 +48,5 @@ jobs:
       uses: codecov/codecov-action@v2
       with:
         files: ./coverage.xml
-        fail_ci_if_error: true
+        fail_ci_if_error: false
```
Comment (joaomcteixeira) [inline .github/workflows/tests.yml:51]: Why change this to `false`?

File: tools/haddock-score.py (hunk 1 above)
Comment (rvhonorato) [reply tools/haddock-score.py]: > air will be in principle 0
> may-be more interesting to output the BSA.
> And I would not output everything by default, but only as an option (e.g. if the `-full` argument is given).
> Simpler for incorporating the score in other scripts.


Added `--full` option


> And same thing for outputting the `pdb` and `psf` files, only if an option is given, e.q. `-outpdb` -`outpsf`


Added `--outputpdb` and `--outputpsf` options

**Comment by rvhonorato** [comment]
> I would rather save it as <original-pdb-filename)_hs.pdb to avoid overwriting it if running on multiple PDBs in the same dir.

Done

**Comment by rvhonorato** [comment]
> I calculated the haddock-score for the first 20 pdbs of 1_emscoring for capri Target218 (check out on tintin if you like). It is highly correlated to the true haddock score, but the pearson coeff. is not super (r~0.88). Should we add more steps to this? Or is it ok as it is?

We figured out that the source of the difference is that @mgiulini was calculating the haddock-scores in the already-minimized models instead of the input ones.

File: MANIFEST.in [hunk 3]
```diff
@@ -7,6 +7,8 @@ include LICENSE
 include requirements.txt
 include requirements.yml
 
+recursive-include tools *.py
```
Comment (joaomcteixeira) [inline MANIFEST.in:10]: I disagree here. I strongly suggest making `haddock-score` a CLI.

File: tools/haddock-score.py [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+    pdb_object : PDBFile
+        The PDBFile object.
+    """
+    model_path = Path(pdb).resolve()
+
+    tidy_model_path = Path(f"{model_path.stem}_tidy.pdb").resolve()
+    with open(model_path, "r") as inp_fh:
+        with open(tidy_model_path, "w") as out_fh:
+            for line in tidy_pdbfile(inp_fh):
+                out_fh.write(line)
```
Comment (joaomcteixeira) [inline tools/haddock-score.py:47]: Single call to save to disk

```suggestion
            lines = list(tidy_pdbfile(inp_fh))
                out_fh.write(os.linesep.join(lines))
```

File: tools/haddock-score.py [hunk 5]
```diff
@@ -0,0 +1,221 @@
+"""
+A simple tool to calculate the HADDOCK-score of a complex.
+
+Usage:
+    python haddock-score complex.pdb
+"""
+
+import argparse
+import subprocess
```
Comment (joaomcteixeira) [inline tools/haddock-score.py:9]:
```suggestion
import os
import subprocess
```

File: tools/haddock-score.py [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+
+    tidy_model_path = Path(f"{model_path.stem}_tidy.pdb").resolve()
+    with open(model_path, "r") as inp_fh:
+        with open(tidy_model_path, "w") as out_fh:
+            for line in tidy_pdbfile(inp_fh):
+                out_fh.write(line)
+
+    main_topoaa_cns_script_as_string = Path(
+        topoaa_module_folder, "cns/generate-topology.cns"
+        ).read_text()
```
Comment (joaomcteixeira) [inline tools/haddock-score.py:51]:
```suggestion
    main_topoaa_cns_script_as_string = Path(
        topoaa_module_folder,
        "cns",
        "generate-topology.cns"
        ).read_text()
```

File: tools/haddock-score.py [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+
+    Returns
+    -------
+    component_dic : dict
+        The dictionary with the energetic components.
+    """
+
+    main_emscoring_cns_script_as_string = Path(
+        emscoring_module_folder, "cns/emscoring.cns"
+        ).read_text()
```
Comment (joaomcteixeira) [inline tools/haddock-score.py:121]:
```suggestion
    main_emscoring_cns_script_as_string = Path(
        emscoring_module_folder,
        "cns",
        "emscoring.cns",
        ).read_text()
```

**Review by joaomcteixeira — COMMENTED** [review]
I just made small code suggestions. I think the `topo` and `emscoring` wrappers are a great opportunity to try to use the module objects directly, or at least to study and learn why we can't yet (if that's the case) use them here already. Anyway, that can be for another PR if there's rush merging this one.

Finally, I really think we should have `haddock-score` as a client as all that is necessary is already placed.

Cheers,

**Comment by joaomcteixeira** [comment]
What is the difference between this script and using `haddock3` with the config: ?

```toml
molecules = "complex.pdb"
run_dir = "whatever"
[topoaa]
[emscoring]
```

**Comment by rvhonorato** [comment]
@joaomcteixeira work on your own changes and merge when its done

**Comment by amjjbonvin** [comment]
It could be both - now it is also used internally in some other design module Rodrigo is working on

**Comment by amjjbonvin** [comment]
Simple one line command that directly returns the score (and nothing else - all in memory unless the options to write out PDB/PSF files are specified).

Actually a feature request from an industrial user. 

And this score can be used in other programs/scripts for optimisation purposes for example.

**Comment by joaomcteixeira** [comment]
Hi Alexandre, thanks for the explanation. Some comments follow:

> Simple one line command that directly returns the score (and nothing else - all in memory unless the options to write out PDB/PSF files are specified

I understand the one-line command. It's a good idea. The "all in memory" is incorrect as the script proposed here writes several files to disk and removes them afterwards. So, if you run this script to thousands of complexes there will be thousands of disk writing/erasing operations, similar to running the normal `haddock3` workflow. To my knowledge, we can't overcome the file writing/reading/erasing because both modules communicate through CNS.

So my question is, is there anything computationally different here from piping the `topoaa` and `emscore`? If not, we can make a CLI that uses the internal features of HADDOCK3 to calculate the score and then retrieve the needed information and files. I am not saying that what's done here is wrong. Simply there's a missed opportunity to use what is already in HADDOCK3 and provide an example of how to use the modules as a library (thinking in the future of jupyter notebooks). I am saying this because I see that this implementation repeats many of the internals that are implemented in the modules already. I am trying to understand the differences between this and the modules and how we can improve the modules if additional features are needed.

But even if we keep the code here presented, moving it to a CLI is needed, IMHO. If there's a hurry now, I can move this to a CLI in this PR and then work on the re-do in another PR, if desired. The concept of "tools" is already present in HADDOCK3 in the form of CLIs.

Cheers,

**Comment by amjjbonvin** [comment]
> I understand the one-line command. It's a good idea. The "all in memory" is incorrect as the script proposed here writes several files to disk and removes them afterwards. So, if you run this script to thousands of complexes there will be thousands of disk writing/erasing operations, similar to running the normal haddock3 workflow. To my knowledge, we can't overcome the file writing/reading/erasing because both modules communicate through CNS.
> 
Ok - clear - but if you are only interesting in the score, then no need to keep them all. Again simple for use for some application (e.g. to use the haddock score in some other program without the need to write config files). Sense and simplicity. 

A simple way to get the HADDOCK score directly from a PDB file is something we have been asked often for in the past. E.g. to calculate features for e.g. AI/ML.
> So my question is, is there anything computationally different here from piping the topoaa and emscore? If not, we can make a CLI that uses the internal features of HADDOCK3 to calculate the score and then retrieve the needed information and files. I am not saying that what's done here is wrong.
> 
Isn’t that what this script does? + erasing the written files.

**Comment by joaomcteixeira** [comment]
Totally agree with you. Simply, the script is rewriting several parts that are already coded in the modules. This means that this script is likely to become useless if something in the modules changes. When I first saw the code, I thought we could use the current HADDOCK3 interfaces more instead of rewriting/copying implementations. Since @rvhonorato assigned me to this PR, I will work on it. If I see my point is not 100% correct, I will keep the current implementation, that's for sure. Because this is a "tool", I imagine there's no rush in merging this PR.

**Comment by joaomcteixeira** [comment]
Updated the scoring script. I am testing now one detail in `topoaa` to ensure everything works. Test it also please to ensure it gives the scores you desire. The `BSA` is not used currently in the scoring function, please check.

Remember to `python setup.py develop` to install the new CLI: `haddock3-score`.

Thoughts for the future: We need to rework a bit on the path managing inside the workflow for the modules to work properly independently, for example, in Jupyter notebooks or to allow higher flexibility in cases like this one. But for now, the implemented strategy uses the available APIs, which I consider a more solid approach.

I will let you know once all tests finish, meanwhile, you can test it.
Cheers,

File: src/haddock/clis/cli_score.py [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+
+    if outputpdb:
+        shutil.copy(
+            Path("run_dir", "1_emscoring", "emscoring_1.pdb"),
+            Path(f"{input_pdb.name}_hs.pdb"),
+            )
+
+    if outputpsf:
+        shutil.copy(
+            Path("run_dir", "0_topoaa", f'{input_pdb.name}_haddock.psf'),
```
Comment (mgiulini) [inline src/haddock/clis/cli_score.py:133]:
```suggestion
            Path(run_dir, "0_topoaa", f'{input_pdb.name}_haddock.psf'),
```

File: src/haddock/clis/cli_score.py [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+        + 0.2 * elec \
+        + 1.0 * desolv \
+        + 0.1 * air
+
+    print("> HADDOCK-score = (1.0 * vdw) + (0.2 * elec) + (1.0 * desolv) + (0.1 * air)")  # noqa: T201, E501
+    print(f"> HADDOCK-score (emscoring) = {haddock_score_itw:.4f}")  # noqa: T201, E501
+
+    if outputpdb:
+        shutil.copy(
+            Path("run_dir", "1_emscoring", "emscoring_1.pdb"),
```
Comment (mgiulini) [inline src/haddock/clis/cli_score.py:127]:
```suggestion
            Path(run_dir, "1_emscoring", "emscoring_1.pdb"),
```

**Review by mgiulini — COMMENTED** [review]
I was testing it and found a mini-bug. I wonder if it makes sense to give the user the possibility to choose the number of energy minimization steps (`nemsteps`) directly from the cli..In my case study it would be useful

**Comment by joaomcteixeira** [comment]
> I was testing it and found a mini-bug. I wonder if it makes sense to give the user the possibility to choose the number of energy minimization steps (nemsteps) directly from the cli..In my case study it would be useful

Yes, this is possible. We can make it possible to pass any `emscoring` parameter. I will make it.

**Comment by mgiulini** [comment]
oh, another thing: maybe when the `outputpdb` and/or `outputpsf` options are specified we could print the name of the file that has been written..what do you think?

**Comment by joaomcteixeira** [comment]
sounds good :+1:

**Comment by joaomcteixeira** [comment]
Notethis is not ready to merge. I made an edit in the `topoaa` I need to confirm it does not affect the workflows. Maybe it still needs some additional work internally.

**Comment by mgiulini** [comment]
> Notethis is not ready to merge. I made an edit in the `topoaa` I need to confirm it does not affect the workflows. Maybe it still needs some additional work internally.

sure! I was testing it and noticed a couple of improvements :)

**Comment by joaomcteixeira** [comment]
@mgiulini can you give a look now, also to the new parameter `-p`? Let me know if the program outputs the values you are expecting. Cheers

**Comment by joaomcteixeira** [comment]
All workflows perform correctly despite the change in `topoaa`. Everything looks good. :+1:

**Comment by rvhonorato** [comment]
> I was testing it and found a mini-bug. I wonder if it makes sense to give the user the possibility to choose the number of energy minimization steps (`nemsteps`) directly from the cli..In my case study it would be useful

Don't think this makes sense, the idea is to calculate the haddock score according to the defined weights, not use the calculation routine to do minimisation steps.

I know it's doable but in my opinion it falls out of scope, sounds like this is the use case for the default workflow pipeline.

**Comment by amjjbonvin** [comment]
There is however a minimization step and I think it is nice to be able to control the number of steps. Even set it to 0 if you really want the score of the “raw” model

**Comment by rvhonorato** [comment]
But then we need to make this very clear, imagine someone is trying to recalculate the haddock score of a given complex and the numbers are different because the minimisation steps were not the default.

Being able to customize the steps is nice indeed, I'm just thinking about reproducibility.

Maybe print it out in big letter that's the steps were altered and that this information needs to accompany the publication.

**Comment by amjjbonvin** [comment]
My guess is most users will simply use the default settings

**Comment by mgiulini** [comment]
Hi there! I agree with @rvhonorato that the use of non-default parameters is dangerous and that this should be made very clear during the execution. I created a couple of warnings for that. If you agree l will give the possibility to modify the weights.

**Comment by joaomcteixeira** [comment]
To help solving the  discussion, the `haddock3-score` CLI will use the default values unless otherwise specified. Currently, users can modify any parameter of the `emscore` using the `-p` option, for example:

```bash
haddock3-score complex.pdb -p nemsteps 50 w_air 1 electflag True
```

See more with `haddock3-score -h`.

For reporting purposes, users can always save the command line use.

I am okay with the warning messages proposed by @mgiulini .

For me, it is okay to be merged. :+1: Good work everyone :rocket:

**Review by joaomcteixeira — APPROVED** [review]

**Comment by mgiulini** [comment]
hello, I changed a bit the code to consider (possibly) new weights in the calculation of the score.

**Comment by joaomcteixeira** [comment]
I cant test or approve now because i am on mobile. Looks great and the
change was necessary because previously if users injected the weights they
were not being used and now they are. Great work!

**Review by joaomcteixeira — APPROVED** [review]
Sounds good @mgiulini :+1:

File: src/haddock/clis/cli_score.py [hunk 10]
```diff
@@ -0,0 +1,270 @@
...
+
+    # config all parameters are correctly spelled.
+    default_emscoring = read_from_yaml_config(DEFAULT_CONFIG)
+    ems_dict = default_emscoring.copy()
+    n_warnings = 0
+    for param in kwargs:
+        if param not in default_emscoring:
+            sys.exit(f'* ERROR * Parameter {param!r} is not a valid `emscoring` parameter')  # noqa:E501
+        if kwargs[param] != default_emscoring[param]:
+            print(f"* WARNING * Value ({kwargs[param]}) of parameter {param} different from default ({default_emscoring[param]})")  # noqa:E501
```
[the remark this answers is not in the document]
Comment (mgiulini) [reply src/haddock/clis/cli_score.py:192]: I understand your point. I will substitute WARNING with ATTENTION

**Pull request merged**